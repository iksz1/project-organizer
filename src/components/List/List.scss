@import "../../vars";

.list {
  flex: none;
  width: 320px;
  margin: 1em;
  will-change: transform, opacity;
  animation: appear 1s ease;
}

.box {
  padding: 0.5em 1em;
  margin-bottom: 0.5em;
  &:first-child {
    margin-top: 0.5em;
  }
  // background: $bg;
  // width: 376px;
  // height: 20vw
  // padding: 2em 1em;
  // background: #cbcebc;
  background: $bg;
  box-shadow: 0 1px 2px 0 rgba(34, 36, 38, 0.15);
  // border: 1px solid #d5d5d5;
  // border-radius: 1em;
  // box-shadow: 1px 1px 5px 0px #b1b1b1;

  // animation-delay: 400ms;
  // animation-fill-mode: backwards;
  // animation-direction: reverse;
}

.header {
  padding: 0.5em 1em;
  background: $bg3;
  color: $textLight;
  // font-weight: 500;
}

// main {
//   padding: 1em 1em;
//   line-height: 1.6;
// }

.footer {
  font-size: small;
  padding: 1em;
  // text-align: center;
  // border-top: 1px solid #d5d5d5;
  background: $bg3;
  color: $textLight;
}

.drag-class {
  border: 2px solid red;
  opacity: 0.5;
}

@keyframes appear {
  0% {
    opacity: 0;
    transform: translateY(-50px);
    // transform: scale(0.1, 0.1, 1);
  }
  100% {
    opacity: 1;
    transform: translateY(0);
    // transform: scale(1, 1, 1);
  }
}
