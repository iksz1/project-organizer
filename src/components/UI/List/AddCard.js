import React, { Component } from "react";
import PropTypes from "prop-types";
import { Button, TextArea } from "../EditForm/EditForm";

export default class AddCard extends Component {
  static propTypes = {
    onSubmit: PropTypes.func
  };

  state = {
    inputValue: ""
  };

  handleTextChange = e => {
    this.setState({ inputValue: e.target.value });
  };

  handleSubmit = e => {
    e.preventDefault();
    this.props.onSubmit(this.state.inputValue);
    this.setState({ inputValue: "" });
  };

  render() {
    const { inputValue } = this.state;

    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <TextArea value={inputValue} onChange={this.handleTextChange} rows="2" />
          {inputValue && <Button type="submit">submit</Button>}
        </form>
      </div>
    );
  }
}
