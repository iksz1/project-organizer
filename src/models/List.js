import { types, flow, destroy } from "mobx-state-tree";
import { Card } from "./Card";
import db from "../utils/dbWrapper";
import calcOrder from "../utils/reorderHelper";

export const List = types
  .model("List", {
    id: types.number,
    boardId: types.number,
    name: types.string,
    cards: types.optional(types.array(Card), []),
    order: types.optional(types.number, 0),
    deleted: types.optional(types.boolean, false)
  })
  .actions(self => ({
    changeName: flow(function* changeName(name) {
      yield db.update("lists", { id: self.id, name });
      self.name = name;
    }),

    addCard: flow(function* addCard(text) {
      const card = yield db.add("cards", {
        text,
        listId: self.id,
        boardId: self.boardId,
        order: calcOrder({ toArr: self.cards })
      });
      self.cards.push(card);
    }),

    deleteCard: flow(function* deleteCard(item) {
      yield db.remove("cards", item.id);
      destroy(item);
    })
  }));
