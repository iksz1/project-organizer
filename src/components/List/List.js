import React, { Component } from "react";
import PropTypes from "prop-types";
import style from "./List.scss";
import Card from "../Card";
import { Container, Draggable } from "react-smooth-dnd";
import AddCard from "../AddCard";
import { observer } from "mobx-react";

// @inject("store")
@observer
class List extends Component {
  static propTypes = {
    list: PropTypes.object.isRequired,
    onListDelete: PropTypes.func,
    onCardMove: PropTypes.func
  };

  onCardDrop = ({ addedIndex, payload }) => {
    const { list, onCardMove } = this.props;
    //same as (index !== null && index !== undefined)
    if (addedIndex != null) {
      const { fromList, fromIndex, card } = payload;
      console.log(`Received from list/index: ${fromList}/${fromIndex}`);
      console.log(`Added to index: ${addedIndex}`);
      // if (fromList === list.id && fromIndex === addedIndex) return; //exit if position hasn't changed
      onCardMove(card, { fromList, fromIndex, toList: list.id, toIndex: addedIndex });
    }
  };

  movingCardPayload = index => {
    const list = this.props.list;
    return { fromList: list.id, fromIndex: index, card: list.cards[index] };
  };

  render() {
    const { list, onListDelete } = this.props;
    // const cards = this.props.store.cards.filter(card => card.listId === list.id);

    return (
      <div className={style.list}>
        <div className={style.header}>
          {list.name}&nbsp;
          <button onClick={() => list.changeName("kek")}>U</button>
          <button onClick={() => onListDelete(list)}>X</button>
        </div>
        <div>
          {/* <ul style={{ margin: 0, padding: 0, listStyle: "none" }}> */}
          <Container
            onDrop={this.onCardDrop}
            getChildPayload={this.movingCardPayload}
            dragClass={style.dragClass}
            groupName="list"
          >
            {list.cards.map(card => (
              <Draggable key={card.id}>
                <Card key={card.id} card={card} onCardDelete={list.deleteCard} />
              </Draggable>
            ))}
          </Container>
          <AddCard handleAction={list.addCard} />
          {/* </ul> */}
        </div>
      </div>
    );
  }
}

export default List;
