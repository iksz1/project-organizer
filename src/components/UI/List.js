import React, { Component } from "react";
import PropTypes from "prop-types";
import { observer } from "mobx-react";
import { Container, Draggable } from "react-smooth-dnd";
import styled from "styled-components";
import Card from "./Card";
import AddCard from "./AddCard";
import EditForm from "./EditForm";
import PopupButtons from "./PopupButtons";

export const Wrapper = styled.div`
  width: 30rem;
  margin: 0.5em;
  padding: 0.5em;
  background: ${props => props.theme.bgList};
  box-shadow: ${props => props.theme.boxShadow};
  border-radius: 0.2em;
  word-wrap: break-word;
`;

export const ListHeader = styled.div`
  position: relative;
  padding: 0.5em 1em;
  margin-bottom: 0.5em;
  text-align: ${props => (props.centered ? "center" : "inherit")};
  background: ${props => props.theme.bgHeader};
  color: ${props => props.theme.textAlt};
  font-weight: 500;
  border-radius: 0.2em;
`;

class List extends Component {
  static propTypes = {
    list: PropTypes.object.isRequired,
    onListDelete: PropTypes.func,
    onCardMove: PropTypes.func
  };

  state = {
    editMode: false
  };

  toggleEditMode = () => {
    this.setState(prevState => ({ editMode: !prevState.editMode }));
  };

  handleUpdate = name => {
    this.props.list.changeName(name);
    this.toggleEditMode();
  };

  onCardDrop = ({ addedIndex, payload }) => {
    const { list, onCardMove } = this.props;
    if (addedIndex != null) {
      const { fromArr, fromIndex, card } = payload;
      onCardMove(card, { fromArr, fromIndex, toArr: list.cards, toIndex: addedIndex });
    }
  };

  movingCardPayload = index => {
    const cards = this.props.list.cards;
    return { fromArr: cards, fromIndex: index, card: cards[index] };
  };

  render() {
    const { list, onListDelete } = this.props;
    const editMode = this.state.editMode;
    const popupItems = [
      { handler: this.toggleEditMode, icon: "edit" },
      { handler: () => onListDelete(list), icon: "trash" }
    ];

    return (
      <Wrapper>
        <ListHeader className={editMode ? "" : "draggable with-popup"}>
          {editMode ? (
            <EditForm
              text={list.name}
              onSubmit={this.handleUpdate}
              onCancel={this.toggleEditMode}
            />
          ) : (
            list.name
          )}
          <PopupButtons items={popupItems} />
        </ListHeader>
        <Container
          onDrop={this.onCardDrop}
          getChildPayload={this.movingCardPayload}
          dragClass="drag-class"
          groupName="cards"
          dragBeginDelay={5}
          dragHandleSelector=".draggable"
          nonDragAreaSelector=".non-draggable"
        >
          {list.cards.map(card => (
            <Draggable key={card.id} style={{ overflow: "inherit" }}>
              <Card key={card.id} card={card} onCardDelete={list.deleteCard} />
            </Draggable>
          ))}
        </Container>
        <AddCard onSubmit={list.addCard} />
      </Wrapper>
    );
  }
}

export default observer(List);
