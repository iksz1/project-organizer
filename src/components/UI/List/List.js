import React, { Component } from "react";
import PropTypes from "prop-types";
import { observer } from "mobx-react";
import Card from "../Card/Card";
import { Container, Draggable } from "react-smooth-dnd";
import styled from "styled-components";
import AddCard from "./AddCard";
import EditForm from "../EditForm/EditForm";
import ItemControls from "../ItemControls/ItemControls";

export const ListWrapper = styled.div`
  width: 300px;
  margin: 0.5em;
  padding: 0.5em;
  background: ${props => props.theme.bgList};
  box-shadow: ${props => props.theme.boxShadow};
  border-radius: 0.2em;
`;

export const ListHeader = styled.div`
  position: relative;
  padding: 0.5em 1em;
  margin-bottom: 0.5em;
  text-align: ${props => (props.centered ? "center" : "inherit")};
  background: ${props => props.theme.bgHeader};
  color: ${props => props.theme.textLight};
  border-radius: 0.2em;
  &:hover .item-controls {
    display: block;
  }
`;

class List extends Component {
  static propTypes = {
    list: PropTypes.object.isRequired,
    onListDelete: PropTypes.func,
    onCardMove: PropTypes.func
  };

  state = {
    editMode: false
  };

  toggleEditMode = () => {
    this.setState(prevState => ({ editMode: !prevState.editMode }));
  };

  handleUpdate = name => {
    this.props.list.changeName(name);
    this.toggleEditMode();
  };

  onCardDrop = ({ addedIndex, payload }) => {
    const { list, onCardMove } = this.props;
    //same as (index !== null && index !== undefined)
    if (addedIndex != null) {
      const { fromArr, fromIndex, card } = payload;
      onCardMove(card, { fromArr, fromIndex, toArr: list.cards, toIndex: addedIndex });
    }
  };

  movingCardPayload = index => {
    const cards = this.props.list.cards;
    return { fromArr: cards, fromIndex: index, card: cards[index] };
  };

  render() {
    const { list, onListDelete } = this.props;
    const editMode = this.state.editMode;

    return (
      <ListWrapper>
        {!editMode && (
          <ListHeader className="list-header">
            {list.name}
            <ItemControls onEdit={this.toggleEditMode} onDelete={() => onListDelete(list)} />
          </ListHeader>
        )}
        {editMode && (
          <ListHeader>
            <EditForm
              text={list.name}
              onSubmit={this.handleUpdate}
              onCancel={this.toggleEditMode}
            />
          </ListHeader>
        )}
        <Container
          onDrop={this.onCardDrop}
          getChildPayload={this.movingCardPayload}
          dragClass="drag-class"
          groupName="cards"
          dragBeginDelay={5}
          dragHandleSelector=".card-wrapper"
          nonDragAreaSelector=".item-controls"
        >
          {list.cards.map(card => (
            <Draggable key={card.id}>
              <Card key={card.id} card={card} onCardDelete={list.deleteCard} />
            </Draggable>
          ))}
        </Container>
        <AddCard onSubmit={list.addCard} />
      </ListWrapper>
    );
  }
}

export default observer(List);
