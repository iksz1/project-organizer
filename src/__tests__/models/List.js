import db from "../../utils/dbWrapper";
import List from "../../stores/models/List";

jest.mock("../../utils/dbWrapper");

const listData = {
  id: 1,
  boardId: 1,
  name: "ideas"
};

it("creates instance successfully", async () => {
  const list = List.create(listData);
  db.update.mockResolvedValue();
  await list.changeName("features");
  expect(list.toJSON()).toMatchSnapshot();
});

it("can add a card", async () => {
  const list = List.create(listData);
  db.add = jest.fn((table, card) => Promise.resolve({ ...card, id: 1 }));
  await list.addCard("add theming");
  expect(list.toJSON()).toMatchSnapshot();
});
