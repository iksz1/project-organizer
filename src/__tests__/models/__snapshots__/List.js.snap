// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add a card 1`] = `
Object {
  "boardId": 1,
  "cards": Array [
    Object {
      "boardId": 1,
      "deleted": 0,
      "id": 1,
      "labelId": 0,
      "listId": 1,
      "order": 1,
      "text": "add theming",
    },
  ],
  "deleted": 0,
  "id": 1,
  "name": "ideas",
  "order": 0,
}
`;

exports[`creates instance successfully 1`] = `
Object {
  "boardId": 1,
  "cards": Array [],
  "deleted": 0,
  "id": 1,
  "name": "features",
  "order": 0,
}
`;
